src/api/v1/minecraft.go

```diff
@@ -0,0 +1,188 @@
+	for key, url := range urls {
+		res, err := http.Get(url)
+		if err != nil {
+			// Hm, error getting one of the urls
+			fmt.Println("Error getting", key, err.Error())
+			continue
+		}
+		if res.StatusCode != http.StatusOK {
+			// wtf
+			fmt.Println("Error getting", key, res.StatusCode)
+			continue
+		}
+
+		body, _ := ioutil.ReadAll(res.Body)
+		lists[key] = strings.Split(string(body), "\n")
+	}
+	return
+}
+
+// Convert a [roleID][]uuid map to a [hashedUUID]role map
+func mapLegacyListsToUserInfoList(lists map[string][]string) (info map[string]*userInfo) {
+	defaults := map[string]userInfo{
+		"developer": {
+			Roles: []role{{ID: "developer", Rank: 0}},
+			Cape:  "http://i.imgur.com/X9NYKct.png",
+		},
+		"staff": {
+			Roles: []role{{ID: "staff", Rank: 2}},
+			Cape:  "http://i.imgur.com/uh6QcuF.png",
+		},
+		"pepsi": {
+			Roles: []role{{ID: "pepsi", Rank: 1}},
+			Icon:  "https://raw.githubusercontent.com/ImpactDevelopment/Resources/master/textures/Pepsi_32.png",
+			Cape:  "http://i.imgur.com/SKjRGbH.png",
+		},
+		"premium": {
+			Roles: []role{{ID: "premium", Rank: 3}},
+			Cape:  "http://i.imgur.com/fc8gsyN.png",
+		},
+	}
+
+	info = make(map[string]*userInfo, sumLists(lists))
+	for key := range lists {
+		for _, line := range lists[key] {
+			// Send a hash of the uuid, not the uuid itself
+			// to make it harder to just bulk-ban users
+			hash := hashUUID(line)
+
+			if _, ok := info[hash]; !ok {
+				// New user, copy the default info for this key
+				defaultInfo := defaults[key]
+				info[hash] = &defaultInfo
+			} else {
+				// Existing user, add roles
+				for _, role := range defaults[key].Roles {
+					err := info[hash].AddRole(role)
+					if err != nil {
+						continue
+					}
+
+					// If this role outranks the others, override capes and icons
+					if info[hash].IsHighest(role) {
+						cape := defaults[key].Cape
+						icon := defaults[key].Icon
+						if cape != "" {
+							info[hash].SetCape(cape)
+						}
+						if icon != "" {
+							info[hash].SetIcon(icon)
+						}
+					}
+				}
+			}
+		}
+	}
+	return
+}
+
+func sumLists(m map[string][]string) (sum int) {
+	sum = 0
+	for key := range m {
+		sum += len(m[key])
+	}
+	return
+}
+
+// True if the role, or a conflicting role, is in the slice
+func containsMatch(s []role, e role) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+		if a.Rank == e.Rank {
+			return true
+		}
+		if a.ID == e.ID {
+			return true
+		}
+	}
+	return false
+}
+
+// Add a role to a userInfo.
+func (info *userInfo) AddRole(r role) error {
+	if containsMatch(info.Roles, r) {
+		return errors.New("userInfo.AddRole: conflicting role already exists")
+	}
+
+	info.Roles = append(info.Roles, r)
+	return nil
+}
+
+func (info *userInfo) SetCape(cape string) {
+	info.Cape = cape
+}
+
+func (info *userInfo) SetIcon(icon string) {
+	info.Icon = icon
+}
+
+func (info userInfo) IsHighest(r role) bool {
+	roles := info.Roles
+	sort.Slice(roles, func(i, j int) bool {
```

Comment:
i don't think this is thread safe but since you construct it for every request it's probably okay...?

Reply:
it is sorting a _copy_ of info.Roles so i think so. Also the info/roles structs are created by the request in the first place so it should all be single threaded.

Reply:
well structs are pass by value in go, but slices are still pass by reference sadly

Reply:
probably won't be an issue, will resolve (maybe comment tho? idk)

Reply:
Oh I see what you mean. You're right it doesn't _need_ to be thread safe, although I'm not sure exactly how it could be made so anyway